### PR TITLE
Fix OD `num_classes` issue.

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
## What has changed and why?

The object detection decoder and postprocessor are not initialized with the actual `num_classes` in the dataset, but use the default. This has resulted in issues like label mismatch when using the model for inference and improper weight loading.

This PR passes the correct `num_classes` in the decoder and postprocessor while handling the weight loading by re-init class-dependent parameters once `num_classes` do not match.

## How has it been tested?

Manually.

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [x] Yes
- [ ] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_task_models/object_detection_components/rtdetrv2_decoder.py [hunk 1]
```diff
@@ -484,6 +486,25 @@ def __init__(
             self.register_buffer("anchors", anchors, persistent=False)
             self.register_buffer("valid_mask", valid_mask, persistent=False)
 
+        if hasattr(self, "register_load_state_dict_pre_hook"):
```
Comment (stegmuel) [inline src/lightly_train/_task_models/object_detection_components/rtdetrv2_decoder.py:493]: Can you add a note (either at the top or in notice.md) about the changes to this file?

File: src/lightly_train/_task_models/task_model_helpers.py [hunk 2]
```diff
@@ -309,6 +340,85 @@ def class_head_reuse_or_reinit_hook(
...
+    state_dict: dict[str, Any],
+    *,
+    weight_key: str,
+    bias_key: str,
+) -> bool:
+    score_head_weight = state_dict.get(weight_key)
+    if score_head_weight is None:
+        return False
+
+    num_classes_state = score_head_weight.shape[0]
```
Comment (stegmuel) [inline src/lightly_train/_task_models/task_model_helpers.py:414]: What about the case where [`query_select_method==agnostic`](https://github.com/lightly-ai/lightly-train/blob/1d8aff2c1671f82dafe20e416e713f546c3a836a/src/lightly_train/_task_models/object_detection_components/rtdetrv2_decoder.py#L465)? Should we leave a TODO and/or a warning?

**Review by stegmuel — COMMENTED** [review]
reviewed

File: src/lightly_train/_task_models/task_model_helpers.py (hunk 2 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/task_model_helpers.py:414]: In this case, the dimension of `enc_score_head` is independent of `num_classes`. With the current impl, we simply load the weights when the dimensions match, and that's consistent with the behavior of other hooks. I think it is fine to load the weights in this case, unless it does not make sense at all or affects the results significantly. 

I don't know the details of how the `score_head` and `query_select_method` work. Should the head be in any case re-inited theoretically speaking when we train on a different dataset?

Comment (stegmuel) [reply src/lightly_train/_task_models/task_model_helpers.py:414]: Let's keep it like this for the moment as it is not the default setting and unlikely to be used.
